update gravity - improve domain validation processing speed

rrobgill opened the pull request:
**What does this PR aim to accomplish?:**

Speed up operation of Update Gravity / `pihole -g`

**How does this PR accomplish the above?:**

Incorporates several small efficiency gains within the main domain validation loop. The individual gains per execution are comparatively minor but as they applied to every entry in blocklists and allowlists during the update process, their effects are cumulative.

- Use a lookup table for valid characters, with single comparison per character instead of repeated comparisons and branches per character

- Short circuit the ip address testing rather than testing every string as an IPv4 and IPv6 address:
   Only perform full ipv4 address test if first character of token is numeric
   Only perform full ipv6 address test if the line contains a colon, 
   and a colon is found within the first 5 characters of the token
 
- Limits check for unicode BOM to once per file

**Test results**

On test system with the following lists:

```
https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts	
https://media.githubusercontent.com/media/zachlagden/Pi-hole-Optimized-Blocklists/refs/heads/main/lists/all_domains.txt	
https://cdn.jsdelivr.net/gh/hagezi/dns-blocklists@latest/adblock/ultimate.txt (ABP style, added as blocklist)
https://raw.githubusercontent.com/arapurayil/aBL/refs/heads/main/filters/main.txt (ABP style - added as allowlist)
https://raw.githubusercontent.com/DandelionSprout/adfilt/refs/heads/master/Alternate%20versions%20Anti-Malware%20List/AntiMalwareDomains.txt 
```

Results of `time pihole -g`

Before (current dev):
real	0m27.348s
user	0m20.069s
sys	0m3.631s

After (this patch):
real	0m22.964s
user	0m15.607s
sys	0m3.598s


**Link documentation PRs if any are needed to support this PR:**

N/A

---
**By submitting this pull request, I confirm the following:**

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against.
2. I have commented my proposed changes within the code and I have tested my changes.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.2)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
6. I have checked that another pull request for this purpose does not exist.
7. I have considered, and confirmed that this submission will be valuable to others.
8. I accept that this submission may not be used, and the pull request closed at the will of the maintainer.
9. I give this submission freely, and claim no ownership to its content.

---
- [x] I have read the above and my PR is ready for review. *Check this box to confirm*

yubiuser reviewed (commented):
I see im improvements around 1-2 secs for my particular lists. I guess it depends on your hardware and list content how much improvement one sees.

DL6ER commented on src/tools/gravity-parseList.c:
```diff
@@ -108,7 +113,7 @@ static inline bool __attribute__((pure)) valid_abp_domain(const char *line, cons
 
 		// First four characters must be "@@||"
-		if(line[0] != '@' || line[1] != '@' || line[2] != '|' || line[3] != '|')
+		if (memcmp(line, "@@||", 4) != 0)
```
Is calling a function with all the related overhead really faster than doing a direct comparison in code that can be highly efficiently implemented by the compiler?

DL6ER commented on src/tools/gravity-parseList.c:
```diff
@@ -270,18 +275,20 @@ int gravity_parseList(const char *infile, const char *outfile, const char *adlis
-			read -= 3;
+			// Clear first_line flag immediately after check
+			first_line = false;
+			if(read >= 3 && memcmp(line, "\xEF\xBB\xBF", 3) == 0)
```
Similar here: I am pretty sure using `memcmp()` actually increases the overhead.

DL6ER reviewed (changes requested):
Thank you for this pull request. I have made a few comments. The main speed increase likely comes from calling `inet_pton()` which is definitely a worthwhile change. Maybe also small bits from the lookup table (but definitely less). However, the two locations where direct comparisons are replaced with `memcmp()` calls should be undone due to the reasons I outlined.

DL6ER commented on src/tools/gravity-parseList.c:
```diff
@@ -342,6 +352,9 @@ int gravity_parseList(const char *infile, const char *outfile, const char *adlis
 
+		// Flag lines containing colons (potential IPv6 addresses)
+		bool line_has_colon = (strchr(line, ':') != NULL);
+
```
Sorry ... I have one more suggestion but then this is definitely good to go in :-)
```suggestion
```

DL6ER commented on src/tools/gravity-parseList.c:
```diff
@@ -352,14 +365,24 @@ int gravity_parseList(const char *infile, const char *outfile, const char *adlis
+			// Ipv6 address
+			// Only test for IPv6 address if line contained a colon
+			// and it's in this token's first 5 characters
+			if (line_has_colon && memchr(token, ':', 5) != NULL)
```
```suggestion
			if (strchr(line, ':') != NULL && memchr(token, ':', 5) != NULL)
```
Avoid `strchr()` - which walks the entire line - to be called for IPv4 address lines at all.

DL6ER replied:
I'm also wondering it this may be a possible place for the code to crash. `memchr` accesses up to 5 bytes inside `token` but couldn't the latter be shorter than that? So `memchr` possibly looks into uninitialized/unallocated memory?

Something like
``` c
memchr(token, ':', min(strlen(token), 5))
```
would be safer, however, `strlen()` alone will already walk the entire line so you could even use a tight loop here:
``` c
static inline bool string_has_within(const char *s, const char character, const int maxlen)
{
    for (int i = 0; i < maxlen && s[i] != '\0'; ++i)
        if (s[i] == character)
            return true;
    return false;
}
```
and then
```suggestion
			if (string_has_within(token, ':', 5))
```
(note: not tested)

rrobgill committed:
update gravity - improve domain validation processing speed

Use lookup table (const) to test for valid characters in domain names

Short circuit the ip address testing:
Only perform full ipv4 address test if first character of token is numeric
Only perform full ipv6 address test if the line contains a colon,
and a colon is found within the first 5 characters of the token

Limit test for unicode BOM to first line of file

Signed-off-by: Rob Gill <rrobgill@protonmail.com>

rrobgill commented:
Tested now with your latest changes.

The strchr() was definitely more expensive than I realised. 

`pihole -g` is an additional 300ms faster on test machine using set of lists above with your changes applied.
